cmd/uzfs_test/uzfs_rebuilding.c

```diff
@@ -0,0 +1,656 @@
+		diff_data += node->len;
+
+		umem_free(node->buf, node->len);
+		umem_free(node, sizeof (*node));
+		mutex_enter(&r_data.mtx);
+	}
+
+	mutex_exit(&r_data.mtx);
+
+	printf("rebuilding finished.. written:%lu, actual written:%lu\n",
+	    diff_data, to_zvol->rebuild_data.rebuild_bytes);
+	umem_free(io_list, sizeof (*io_list));
+	mutex_destroy(&r_data.mtx);
+	cv_destroy(&r_data.cv);
+
+	uzfs_zvol_set_status(to_zvol, ZVOL_STATUS_HEALTHY);
+
+	mutex_enter(&r_info->mtx);
+	r_info->active = B_FALSE;
+	cv_signal(&r_info->cv);
+	mutex_exit(&r_info->mtx);
+
+	zk_thread_exit();
+}
+
+void
+replica_writer_thread(void *arg)
+{
+	worker_args_t *warg = (worker_args_t *)arg;
+	char *buf[15];
+	int idx, j, err;
+	uint64_t blk_offset, offset, vol_blocks, iops = 0;
+	hrtime_t end, now, replica_down_time, replica_start_time;
+	boolean_t replica_active = B_TRUE;
+	kmutex_t *mtx = warg->mtx;
+	kcondvar_t *cv = warg->cv;
+	int *threads_done = warg->threads_done;
+	uint64_t vol_size = warg->active_size;
+	uint64_t block_size = warg->io_block_size;
+	static uint64_t io_num = 0;
+	boolean_t rebuilding_started = B_FALSE;
+	struct rebuilding_info rebuild_info;
+	kthread_t *rebuilding_thread = NULL;
+	uint64_t mismatch_count = 0;
+
+	for (j = 0; j < 15; j++)
+		buf[j] = (char *)umem_alloc(sizeof (char)*(j+1)*block_size,
+		    UMEM_NOFAIL);
+
+	now = gethrtime();
+	end = now + (hrtime_t)(total_time_in_sec * (hrtime_t)(NANOSEC));
+	replica_start_time = now + (hrtime_t)(total_time_in_sec *
+	    (hrtime_t)(NANOSEC)) - (hrtime_t)(total_time_in_sec/8 *
+	    (hrtime_t)(NANOSEC));
+	replica_down_time = now + (hrtime_t)(total_time_in_sec/4 *
+	    (hrtime_t)(NANOSEC));
+
+	rebuild_info.to_zvol = zvol2;
+	rebuild_info.from_zvol = zvol1;
+	mutex_init(&rebuild_info.mtx, NULL, MUTEX_DEFAULT, NULL);
+	cv_init(&rebuild_info.cv, NULL, CV_DEFAULT, NULL);
+
+	vol_blocks = (vol_size) / block_size;
+
+	if (silent == 0)
+		printf("Starting write..\n");
+
+	while (1) {
+		io_num++;
+		blk_offset = uzfs_random(vol_blocks - 16);
+		offset = blk_offset * block_size;
+
+		idx = uzfs_random(15);
+
+		populate_string(buf[idx], (idx + 1) * block_size);
+
+		err = uzfs_write_data(zvol1, buf[idx], offset,
+		    (idx + 1) * block_size, NULL, B_FALSE);
+		if (err != 0)
+			printf("IO error at offset: %lu len: %lu\n", offset,
+			    (idx + 1) * block_size);
+
+		if (replica_active) {
+			err = uzfs_write_data(zvol2, buf[idx], offset,
+			    (idx + 1) * block_size, NULL, B_FALSE);
+			if (err != 0)
+				printf("IO error at offset: %lu len: %lu\n",
+				    offset, (idx + 1) * block_size);
+		} else {
+			mismatch_count += (idx + 1) * block_size;
+		}
+
+		if (err)
+			exit(2);
+
+		iops += (idx + 1);
+		now = gethrtime();
+		if (now > replica_down_time && now < replica_start_time) {
+			replica_active = B_FALSE;
+		} else if (now > replica_start_time) {
+			replica_active = B_TRUE;
+			if (!rebuilding_started) {
+				mutex_enter(&rebuild_info.mtx);
+
+				rebuilding_thread = zk_thread_create(NULL, 0,
```

Comment:
replica_start_time and replica_down_time are good.. we can have rebuild_start_time also similarly.. we shouldn't assume that rebuild will start at the same time when replica comes up

Reply:
ok